src/ngraph/frontend/onnxifi/backend.cpp

```diff
@@ -0,0 +1,295 @@
+                }
+                else if (m_type == "NNP")
+                {
+                    *reinterpret_cast<::onnxEnum*>(infoValue) = ONNXIFI_DEVICE_TYPE_NPU;
+                }
+                else
+                {
+                    *reinterpret_cast<::onnxEnum*>(infoValue) = ONNXIFI_DEVICE_TYPE_HETEROGENEOUS;
+                }
+            }
+        }
+
+        void Backend::get_extensions(void* infoValue, std::size_t* infoValueSize) const
+        {
+            if (infoValueSize == nullptr)
+            {
+                throw std::invalid_argument{"null pointer"};
+            }
+            *infoValueSize = 0;
+            if (infoValue != nullptr)
+            {
+                *reinterpret_cast<char*>(infoValue) = '\0';
+            }
+        }
+
+        void Backend::get_graph_init_properties(void* infoValue, std::size_t* infoValueSize) const
+        {
+            if (infoValueSize == nullptr)
+            {
+                throw std::invalid_argument{"null pointer"};
+            }
+            *infoValueSize = 0;
+            if (infoValue != nullptr)
+            {
+                *reinterpret_cast<char*>(infoValue) = '\0';
+            }
+        }
+
+        void Backend::get_onnxifi_version(void* infoValue, std::size_t* infoValueSize) const
+        {
+            if (infoValueSize == nullptr)
+            {
+                throw std::invalid_argument{"null pointer"};
+            }
+            std::size_t requested{*infoValueSize};
+            *infoValueSize = sizeof(uint64_t);
+            if (requested < *infoValueSize)
+            {
+                throw std::length_error{"not enough space"};
+            }
+            if (infoValue != nullptr)
+            {
+                *reinterpret_cast<uint64_t*>(infoValue) = 1;
+            }
+        }
+
+        void Backend::get_name(void* infoValue, std::size_t* infoValueSize) const
+        {
+            const constexpr char name[] = "ngraph:";
+            if (infoValueSize == nullptr)
+            {
+                throw std::invalid_argument{"null pointer"};
+            }
+            std::size_t requested{*infoValueSize};
+            *infoValueSize = sizeof name + m_type.size();
+            if (requested < *infoValueSize)
+            {
+                throw std::length_error{"not enough space"};
+            }
+            if (infoValue != nullptr)
+            {
+                std::memcpy(infoValue, name, sizeof name);
+                std::memcpy(
+                    reinterpret_cast<char*>(infoValue) + sizeof name, m_type.data(), m_type.size());
+            }
+        }
+
+        void Backend::get_vendor(void* infoValue, std::size_t* infoValueSize) const
+        {
+            const constexpr char vendor[] = "Intel Corporation";
+            if (infoValueSize == nullptr)
+            {
+                throw std::invalid_argument{"null pointer"};
```

Comment:
The type of exception is important. The message is not going to be presented in any form. It could be the empty string. `std::invalid_argument` will translate to ONNXIFI_STATUS_INVALID_POINTER error message. This error message has the precise meaning described in the ONNXIFI documentation. The ONNXIFI is C API, not C++.